### ISSUE TITLE
Rails testing question

### ISSUE THREAD

**Issue by ThomasBush** [WestMichiganRubyTraining/discussion#53]
I am trying to write tests around the CMS application that I built for work using rspec and factory girl.  My CMS contains many products which can belong to one or many sites.  I have a method which checks for the current site and is causing all the controller specs to fail.  

I figure this is because I don't have a site in my test database, so I began trying to figure out factory girl to create stubbed out users and sites.  This is where I get a little lost.  I created my `spec/factories` folder and populated it with `users.rb` and `sites.rb` found here: https://gist.github.com/ThomasBush/521db20ba0d95bf297b2

I try to use factory girl in my controller specs like this:

```
before (:each) do
    @admin = FactoryGirl.create(:admin)
    sign_in @admin
    @site = FactoryGirl.create(:site)
end
```

For clarity, my intention with the above code is that factory girl would create both a user and site based on the associated ruby files in the `spec/factories` folder. As well as sign in the devise user before running each spec.

This doesn't seem to be working.  I am very new to testing and so I know very little about testing as well as factory girl I am not sure really how to move forward.  I was hoping someone could point me in the right direction.  Is this even the correct approach to solve me problem?  Would these be more appropriately placed in a like the spec_helper or something? As the user will need to be logged in to use the app and there will always need to be at least one site. 

Thanks for any help you can provide!

**Comment by billgathen** [comment]
Can you drop the output of running your tests into a gist and link to it? That would give us more to go on.

**Comment by ThomasBush** [comment]
@billgathen I added the output of one section of my controller specs as there would be so much if I included all controller specs.  So I ran rspec spec/controllers/technologies_spec.rb to arrive at the output located in the same gist here https://gist.github.com/ThomasBush/521db20ba0d95bf297b2

**Comment by billgathen** [comment]
To establish whether the problem is with factory girl, do `p @admin` in your before clause, right after the variable is created. This will tell you if there's anything inside it. (`p @admin` is functionally-equivalent to saying `puts @admin.inspect`, but a lot shorter)

Something to notice is that every single test blows up on line 46 in application_controller.rb. Can you add that method to your gist? Or the whole controller, if it's a reasonable size.

If the @admin data is there, then you may be having issues where the tests are inside a transaction and the process running the controller is outside it: someone can correct me if I've got this wrong. If that's the case, the admin database entry you're creating is invisible to the app itself.

**Comment by ThomasBush** [comment]
I updated the Gist with the correct output.  I also noticed this particular file lacked the

```
before (:each) do
    @admin = FactoryGirl.create(:admin)
    p @admin
    sign_in @admin
    @site = FactoryGirl.create(:site)
  end
```

So the error output has changed as a result of adding this.  Also it appears in the output that I am getting a correct user.  I created a `spec/factories/roles.rb` and `spec/factories/permissions.rb` and included these in the gist as well.  I think looking at the error that I need to connect the user with the proper roles and the role with the proper permissions.  

I think this relates to how I set up my spec/factories/user.rb specifically:

```
factory :admin do
      admin = true
      role_id = 1
    end

    factory :marketing do
      role_id = 4
    end

    factory :apiuser do
      role_id = 8
    end
```

I think I need to replace `role_id = 8` with the factory girl syntax for a relationship. as these ids cant possible be accurate.

**Comment by billgathen** [comment]
You are correct: here's a [link to establishing associations between factories](https://github.com/thoughtbot/factory_girl/blob/master/GETTING_STARTED.md#associations)

You're getting there: keep going!

**Comment by thejbsmith** [comment]
Looking at the errors you are receiving, it looks like you're having an error where the role is not being found. Looking at your load_permissions method in your application controller, `@current.role.permissions...` would throw that error if role is nil. If you set up the associations between factories that will probably take care of the error.

Although you are setting the role_id for each admin, are you sure that role is actually being created? (If not it would lead to the exact error that you are seeing). Adding in the associations like above will ensure that the child and parent objects are both created for the tests.

**Comment by ThomasBush** [comment]
Thanks for the link @billgathen and the comments @thejbsmith.  Alright starting to make a bit more sense and I got 6 of the 16 test to pass! Updated nearly everything in the gist including the output.  I am confused by my errors now, I have no idea what "http://test.host/" is, is this something set up for me automatically?  Also I can perform all basic CRUD functionality on technologies section, so I don't understand why the database persistance tests fail

**Comment by ThomasBush** [comment]
So to update this, I commented out `config.fixture_path = "#{::Rails.root}/spec/fixtures"` from my `spec_helper.rb` file.  I am using FactoryGirl in place of fixtures as mentioned and did not notice this line including fixture files.  This removed quite a few errors so I wanted to include for others who are also new to testing in rails with factory girl and rspec.  

Now I am confused how FactoryGirl integrates with Rspec's valid_attributes method.  This seems to be a duplication to me, is this correct?  For example, my "Role" spec has the following valid_attributes 

```
def valid_attributes
    { "title" => "Super Admin" }
  end
```

This causes an error in that FactoryGirl is creating an instance of Role, and valid_attributes is also creating an instance of Role.  So my test expects to see the one instance in the array, but receives two instance in the array.   I understand why this is an error, but have had difficulty figuring out how to solve.  Should I remove valid_attributes?  Than how do I have Factory girl replace valid_attributes for the creation of the role instance?  Please let me know if this makes sense.  This topic is still new to me so I don't know if I am wording this properly to get at the question I want to know the answer to.

 If I run my entire testing suite I have this error occur a bunch of times.  So solving this specific instance will help me clear up a bunch more errors.

I have updated the output of running this specific test in my gist https://gist.github.com/ThomasBush/521db20ba0d95bf297b2

**Comment by thejbsmith** [comment]
I have never used the valid_attributes method while testing Rspec with FactoryGirl and have never had any issues. I always make sure my default factory for each model contains valid attributes for testing, and even throw in a simple spec testing that the default factory is valid (saves time debugging other tests if I add in a new attribute and see that spec failing). 

I don't think you need to have FactoryGirl replace the valid_attributes for the creation of the instance, I think you just want to use factory girl to create / build each instance.

Hope this helps, if not, can you add in the specific spec file you are looking at so we can look at it?

**Comment by ThomasBush** [comment]
Sorry @thejbsmith, I am still a bit confused.  One note that I should probably point out is that my app requires a user to be logged in to see anything.  So there must be a user, permission and role set before every test.

My method for solving this was to place the below code in the spec_helper.  This way it is in one place and runs before each test. 

```
  config.before(:each, :type => :controller) do
    @admin = FactoryGirl.create(:admin)
    sign_in @admin
    @site = FactoryGirl.create(:site)
  end
```

Than through associations I have `:admin` creates the correct role which in turn creates the correct permissions.  

I have included my roles_controller_spec in the gist.  After reading your comments and using the very first test as an example I think I need to replace `role = Role.create! valid_attributes` with some type of factory girl call, or remove it entirely since this is always created for every test.  But my confusion cone here: 
`assigns(:roles).should eq([role])` 

I don't understand how to use factory girl in the comparison here.  I have tried replacing that line with 
`assigns(:roles).should eq([FactoryGirl.create(:super_role)])`
This fails as the expected and received values have different id since I am creating in the test as well.

**Comment by thejbsmith** [comment]
You can replace `role = Role.create! valid_attributes` with `role = FactoryGirl.create(:role)` if you want to move completely to FactoryGirl, but I don't think that is your problem.

On your GET #index test, the error you are getting is because 2 roles have been created (1 by that test, and another created by your config.before block in your spec_helper). That's why you get the error that the 'expected' only has one role and the 'got' has 2 roles.

There are a few ways to solve this issue. If you change your test from this:

``` ruby
describe "GET index" do
  it "assigns all roles as @roles" do
    role = Role.create! valid_attributes
    get :index, {}
    assigns(:roles).should eq([role])
  end
end
```

to 

``` ruby
describe "GET index" do
  it "assigns all roles as @roles" do
    # role = Role.create! valid_attributes
    get :index, {}
    assigns(:roles).should eq([@admin.role])
  end
end
```

This way you know your app is working properly by only having 1 version of a certain role created.

Another way to handle this would be keep the test the same as you currently have it, and change:

``` ruby
 assigns(:roles).should eq([role])
```

to

``` ruby
 assigns(:roles).should include(role)
```

This should technically get rid of the error, but it really doesn't work the way your app would (since assigns(:role) would still have 2 "Super Admin" roles in it.

**Comment by ThomasBush** [comment]
Thanks @thejbsmith, thats great! I figured out code identical to your first method of solving the problem, came here to post about it, and ask if it was correct.  Your post didn't show up till I published mine, so I removed it.  I really appreciate the help!  I think I owe you another beer from Perrin...
